### PR TITLE
[buddy] Add HPA Support for Proxy in Teleport-Cluster Helm Chart

### PR DESCRIPTION
**Pull request by tigrato**
This PR introduces Horizontal Pod Autoscaler (HPA) support for the proxy deployment in the Teleport-Cluster Helm chart, allowing for automatic scaling based on CPU & Memory utilization. 

Changes:
- Created proxy/hpa.yaml to define the HPA Kubernetes resource & HPA configuration options in values.yaml.
- Updated values.schema.json to validate for the new proxy-hpa configuration.
- Added linter and unit test examples for HPA settings.

Changelog: Ability to create Horizontal Pod Autoscaler for Teleport Proxy using the `teleport-cluster` Helm chart.

---

Buddies #39236

### CODE REVIEW

File: examples/chart/teleport-cluster/.lint/proxy-hpa.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+clusterName: test-cluster
+chartMode: aws
+aws:
+  region: us-west-2
+  backendTable: test-dynamodb-backend-table
+  auditLogTable: test-dynamodb-auditlog-table
+  sessionRecordingBucket: test-s3-session-storage-bucket
+proxy:
+  teleportConfig: {}
```
Comment (webvictim) [inline examples/chart/teleport-cluster/.lint/proxy-hpa.yaml:9]:
```suggestion
```

I don't think this value is required when nothing is being overridden.

File: examples/chart/teleport-cluster/values.schema.json [hunk 2]
```diff
@@ -49,7 +49,39 @@
         },
         "proxy": {
             "$id": "#/properties/proxy",
-            "type": "object"
+            "type": "object",
+            "properties": {
+                "teleportConfig": {
+                    "type": "object"
+                },
```
Comment (webvictim) [inline examples/chart/teleport-cluster/values.schema.json:56]:
```suggestion
```
I don't think this is required for the PR and may cause unrelated issues - probably best to remove it. We don't have the same value under `auth`.

**Review by webvictim — COMMENTED** [review]
The YAML/templating/lint LGTM but I'm not very familiar with HPA.

File: examples/chart/teleport-cluster/.lint/proxy-hpa.yaml (hunk 1 above)
Comment (ssyno) [reply examples/chart/teleport-cluster/.lint/proxy-hpa.yaml:9]: I think we can remove `teleportConfig: {}` as its not required

File: examples/chart/teleport-cluster/values.schema.json (hunk 2 above)
Comment (ssyno) [reply examples/chart/teleport-cluster/values.schema.json:56]: I saw `teleportConfig` exists on the current values, thats why I added it on [values.schema.json](https://github.com/gravitational/teleport/pull/39290/files/206351dfbc134bcfaf2eb214f38a673b5dafa251#diff-d3656ccc8503ae6ef6717561811a6cead6654b8b2f5d79692b6cc68ebf982f9b), we can remove it as you suggest.

**Review by hugoShaka — CHANGES_REQUESTED** [review]
I don't think deploying an HPA for the Teleport proxy service is safe or desirable.

Due to how agents brute-force their tunnel connections through a round-robin LB, it takes several minutes for an agent to connect to a new proxy. A new proxy can cause up to 10 minutes of downtime; this is not theoretical; we've got customers opening support tickets about this.

Downscaling events are also highly disruptive as they will break all ongoing sessions. Session resumption is only supported for SSH in node-recording mode. The current proxy grace period does not allow the resumable SSH sessions to be seamlessly steered to another proxy.

The proxy must be considered stateful because of the persistent connections. This is a persistent connection broker. Users don't think about this, they just enable HPAs to check the box "this will scale" (it won't). Autoscaling is a financial optimization to avoid running at full capacity and loosing money. It won't allow you to support a higher load (in fact it's the opposite due to the delay between when you need the capacity and when the capacity is provided)

Tuning HPAs is difficult and requires load tests and a lot of Kubernetes maturity. We cannot prevent users from manually crafting an HPA and shooting themselves in the foot, but the chart should not encourage them to do so.

**Comment by espadolini** [comment]
Except in very rare cases and in configurations that are not currently handled by the helm chart to begin with (i.e. proxy peering and very large clusters) each proxy will split the load from client connections but will have the exact same static load from agent connectivity; running multiple proxies is for HA, not for scaling.

**Comment by ssyno** [comment]
Thanks a lot for sharing your concerns about adding HPA to the Teleport proxy. It's clear there are some real challenges with connection stability and session resumption during scaling events. Given all mentioned above, it makes sense to drop this HPA attempt as it doesn't align well with the project goals.

**Comment by webvictim** [comment]
Agreed, I'm going to close this PR and the original.

**Pull request closed without merging**